neqo-transport/src/connection/tests/stream.rs

```diff
@@ -111,6 +113,121 @@ fn transfer() {
     assert!(fin3);
 }
 
+#[cfg(test)]
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+struct IdEntry {
+    sendorder: StreamOrder,
+    stream_id: StreamId,
+}
+
+#[cfg(test)]
+// tests stream sendorder priorization
+fn sendorder_test(order_of_sendorder: &Vec<StreamOrder>) {
+    let mut client = default_client();
+    let mut server = default_server();
+    connect_force_idle(&mut client, &mut server);
+
+    qdebug!("---- client sends");
+    // open all streams and set the sendorders
+    let mut ordered = BinaryHeap::new();
+    let mut streams = Vec::<StreamId>::new();
+    for sendorder in order_of_sendorder {
+	streams.push(client.stream_create(StreamType::UniDi).unwrap());
+	let id = *(streams.last()).unwrap();
+	ordered.push(IdEntry { sendorder: *sendorder, stream_id: id });
+	client.streams.set_sendorder(id, sendorder.sendorder).ok();
+    }
+    // Write some data to all the streams
+    for stream_id in streams {
+	client.stream_send(stream_id, &[6; 100]).unwrap();
+    }
+
+    // Sending this much takes a few datagrams.
+    let mut datagrams = vec![];
+    let mut out = client.process_output(now());
+    while let Some(d) = out.dgram() {
+        datagrams.push(d);
+        out = client.process_output(now());
+    }
+    assert_eq!(*client.state(), State::Confirmed);
+
+    qdebug!("---- server receives");
+    for (d_num, d) in datagrams.into_iter().enumerate() {
+        let out = server.process(Some(d), now());
+        assert_eq!(
+            out.as_dgram_ref().is_some(),
+            (d_num + 1) % usize::try_from(DEFAULT_ACK_PACKET_TOLERANCE + 1).unwrap() == 0
+        );
+        qdebug!("Output={:0x?}", out.as_dgram_ref());
+    }
+    assert_eq!(*server.state(), State::Confirmed);
+
+    let mut stream_ids = server.events().filter_map(|evt| match evt {
+        ConnectionEvent::RecvStreamReadable { stream_id, .. } => Some(stream_id),
+        _ => None,
+    });
+
+    // streams should arrive in priority order, not order of creation, if sendorder prioritization
+    // is working correctly
+    let mut recv_streams = Vec::<StreamId>::new();
+    let mut should_recv_streams =  Vec::<StreamId>::new();
+    // walk the ordered list of streams, and handle random ordering within a sendorder
+    let mut last_sendorder: StreamOrder = StreamOrder{sendorder: None};
+    // BinaryHeap's sorted iterator is nightly only for now
+    // the ordering is inverted, however!
+    let mut ordered_vec = ordered.into_sorted_vec();
+    ordered_vec.reverse();
```

Comment:
It seems like maybe you could avoid using BinaryHeap here and just rely on `<&mut []>::sort` or `<&mut []>::sort_by` instead.  (I think that you will find that `Option` implements `Ord` in a way that will fit nicely here.)